third_party/grpc_java/java_grpc_library.bzl

```diff
@@ -0,0 +1,132 @@
+  includes = [f for dep in ctx.attr.srcs for f in dep.proto.transitive_imports]
+
+  flavor = ctx.attr.flavor
+  if flavor == "normal":
+    flavor = ""
+  ctx.action(
+      inputs = [ctx.executable._java_plugin] + srcs + includes,
+      outputs = [srcdotjar],
+      executable = ctx.executable._protoc,
+      arguments = [
+          "--plugin=protoc-gen-grpc-java=" + ctx.executable._java_plugin.path,
+          "--grpc-java_out={0},enable_deprecated={1}:{2}"
+            .format(flavor, str(ctx.attr.enable_deprecated).lower(), srcdotjar.path)]
+          + ["-I{0}={1}".format(_path_ignoring_repository(include), include.path) for include in includes]
+          + [_path_ignoring_repository(src) for src in srcs])
+  ctx.action(
+      command = "cp $1 $2",
+      inputs = [srcdotjar],
+      outputs = [ctx.outputs.srcjar],
+      arguments = [srcdotjar.path, ctx.outputs.srcjar.path])
+
+_gensource = rule(
+    attrs = {
+        "srcs": attr.label_list(
+            mandatory = True,
+            non_empty = True,
+            providers = ["proto"],
+        ),
+        "flavor": attr.string(
+            values = [
+                "normal",
+                "lite",  # Not currently supported
+            ],
+            default = "normal",
+        ),
+        "enable_deprecated": attr.bool(
+            default = False,
+        ),
+        "_protoc": attr.label(
+            default = Label("@com_google_protobuf//:protoc"),
+            executable = True,
+            cfg = "host",
+        ),
+        "_java_plugin": attr.label(
+            default = Label("@grpc_java//compiler:grpc_java_plugin"),
+            executable = True,
+            cfg = "host",
+        ),
+    },
+    outputs = {
+        "srcjar": "%{name}.srcjar",
+    },
+    implementation = _gensource_impl,
+)
+
+def java_grpc_library(name, srcs, deps, flavor=None,
+                      enable_deprecated=None, visibility=None,
+                      **kwargs):
+  """Generates and compiles gRPC Java sources for services defined in a proto
+  file. This rule is compatible with java_proto_library and java_lite_proto_library.
+
+  Do note that this rule only scans through the proto file for RPC services. It
+  does not generate Java classes for proto messages. You will need a separate
+  java_proto_library or java_lite_proto_library rule.
+
+  Args:
+    name: (str) A unique name for this rule. Required.
+    srcs: (list) a single proto_library target that contains the schema of the
+        service. Required.
+    deps: (list) a single java_proto_library target for the proto_library in
+        srcs.  Required.
+    flavor: (str) "normal" (default) for normal proto runtime. "lite"
+        for the lite runtime.
+    visibility: (list) the visibility list
+    **kwargs: Passed through to generated targets
+  """
+  if flavor == None:
+    flavor = "normal"
+
+  if len(deps) > 1:
+    print("Multiple values in 'deps' is deprecated in " + name)
```

Comment:
The upgrade to this implementation of java_grpc_library generates the following DEBUG message indicating deprecation from `BUILD.googleapis`:

DEBUG: $repo_root/bazel-buildfarm/third_party/grpc_java/java_grpc_library.bzl:101:5: Multiple values in 'deps' is deprecated in google_devtools_remoteexecution_v1test_remote_execution_java_grpc

Would prefer to not regress like this in this explicit upgrade.

Reply:
I see that we cannot simply remove the longrunning operations dependency due to references within the remote execution protobuf definition, but I'm confused as to why the grpc authors believe this to be a valid mechanism to deprecate.